core/drivers/scmi-msg/perf_domain.c

```diff
@@ -0,0 +1,407 @@
+}
+
+int32_t __weak plat_scmi_perf_level_get(unsigned int channel_id __unused,
+					unsigned int domain_id __unused,
+					unsigned int *level __unused)
+{
+	return SCMI_NOT_SUPPORTED;
+}
+
+int32_t __weak plat_scmi_perf_level_set(unsigned int channel_id __unused,
+					unsigned int domain_id __unused,
+					unsigned int level __unused)
+{
+	return SCMI_NOT_SUPPORTED;
+}
+
+static void protocol_version(struct scmi_msg *msg)
+{
+	struct scmi_protocol_version_p2a return_values = {
+		.status = SCMI_SUCCESS,
+		.version = SCMI_PROTOCOL_VERSION_PERF_DOMAIN,
+	};
+
+	VERBOSE_MSG("SCMI perf %#"PRIx32, SCMI_PROTOCOL_VERSION_PERF_DOMAIN);
+
+	if (msg->in_size) {
+		scmi_status_response(msg, SCMI_PROTOCOL_ERROR);
+		return;
+	}
+
+	scmi_write_response(msg, &return_values, sizeof(return_values));
+}
+
+static void protocol_attributes(struct scmi_msg *msg)
+{
+	unsigned int channel_id = msg->channel_id;
+	size_t count = plat_scmi_perf_count(channel_id);
+	uint32_t power_in_mw = 0;
+	struct scmi_perf_protocol_attributes_p2a return_values = {
+		.status = SCMI_SUCCESS,
+		.attributes = SCMI_PERF_PROTOCOL_ATTRIBUTES(power_in_mw, count),
+	};
+	void *stats_buf = NULL;
+	size_t stats_len = 0;
+
+	VERBOSE_MSG("channel %u: %zu performance domains", channel_id, count);
+
+	if (msg->in_size) {
+		scmi_status_response(msg, SCMI_PROTOCOL_ERROR);
+		return;
+	}
+
+	stats_buf = plat_scmi_perf_statistics_buf(channel_id, &stats_len);
+	if (stats_len) {
+		paddr_t stats_pa = virt_to_phys(stats_buf);
+
+		if (stats_pa && tee_vbuf_is_non_sec(stats_buf, stats_len)) {
+			return_values.statistics_len = stats_len;
+			reg_pair_from_64((uint64_t)stats_pa,
+					 &return_values.statistics_address_high,
+					 &return_values.statistics_address_low);
+		} else {
+			IMSG("Disable SCMI perf statistics: invalid buffer");
+			DMSG("Stats buffer va %p, pa %#"PRIxPA", size %zu",
+			     stats_buf, stats_pa, stats_len);
+		}
+	}
+
+	scmi_write_response(msg, &return_values, sizeof(return_values));
+}
+
+static void protocol_message_attributes(struct scmi_msg *msg)
+{
+	struct scmi_protocol_message_attributes_a2p *in_args = (void *)msg->in;
+	struct scmi_protocol_message_attributes_p2a return_values = {
+		.status = SCMI_SUCCESS,
+		/* For this protocol, attributes shall be zero */
+		.attributes = 0,
+	};
+
+	if (msg->in_size != sizeof(*in_args)) {
+		scmi_status_response(msg, SCMI_PROTOCOL_ERROR);
+		return;
+	}
+
+	if (!message_id_is_supported(in_args->message_id)) {
+		scmi_status_response(msg, SCMI_NOT_FOUND);
+		return;
+	}
+
+	scmi_write_response(msg, &return_values, sizeof(return_values));
+}
+
+static int32_t sanitize_message(struct scmi_msg *msg, unsigned int *domain_id,
+				size_t exp_in_size)
+{
+	size_t domain_count = plat_scmi_perf_count(msg->channel_id);
+
+	if (msg->in_size != exp_in_size)
+		return SCMI_PROTOCOL_ERROR;
+
+	if (*domain_id >= domain_count)
+		return SCMI_INVALID_PARAMETERS;
+
+	*domain_id = confine_array_index(*domain_id, domain_count);
```

Comment:
I missed this. For index to be properly confined regarding speculative execution once this function returns, it should be unconditionally confined. I mean here that this instruction should be call before this function may return:
```patch
 {
 	size_t domain_count = plat_scmi_perf_count(msg->channel_id);
 
+	*domain_id = confine_array_index(*domain_id, domain_count);
+
 	if (msg->in_size != exp_in_size)
 		return SCMI_PROTOCOL_ERROR;
 
 	if (*domain_id >= domain_count)
 		return SCMI_INVALID_PARAMETERS;
 
-	*domain_id = confine_array_index(*domain_id, domain_count);
-
 	return SCMI_SUCCESS;
 }